pkg/service/service.go

```diff
@@ -1369,6 +1374,35 @@ func (s *Service) restoreBackendsLocked() error {
 			logfields.BackendState:     b.State,
 			logfields.BackendPreferred: b.Preferred,
 		}).Debug("Restoring backend")
+		if _, ok := svcBackendsById[b.ID]; !ok && s.backendRefCount[b.L3n4Addr.Hash()] != 0 {
```

Comment:
@aditighag Too late to the party, but a few questions:
* Isn't the `s.backendRefCount[b.L3n4Addr.Hash()] != 0` check redundant, as the refcount of such a backend can be incremented only if it belongs to a service (thus, `!ok` would evaluate to false).
* Do I read it correctly that backends in the `Terminating` state would be removed before the grace period happens? I think it's not a big deal, as cilium-agent restarts do not expect to happen very often. Just what happens if we receive an EndpointSlice delete event, and cilium-agent doesn't have any reference to previously `Terminating` backends?

Reply:
Another thing it would be helpful if you could elaborate on (e.g., in a commit msg) under what circumstances the backend leaks happened.